# Week 2 Checkpoint: First Data Profile
**BAN 6003 Data Management and Analytics Integration**

Complete this notebook using the `employee_data.csv` file in the `data/` folder.

This is the independent checkpoint for Week 2. Use the functions practiced in the guided lab. You are not expected to clean the data yet. Your goal is to profile the data and report what you notice.


## Instructions

1. Run each cell from top to bottom.
2. Fill in all cells marked `# Your code here`.
3. Write short answers in the markdown response areas.
4. Save, commit, and push your notebook to GitHub.

You may use the Week 2 guided lab as a reference.


## Task 0: Import Pandas and Load the Data
Run the cell below.


In [1]:
import pandas as pd

df = pd.read_csv("../data/employee_data.csv")
df.head()


,employee_id,first_name,last_name,hire_date,department,salary,age,full_time
0,101,John,Smith,4/15/22,Marketing,85000,34.0,True
1,102,Jane,Doe,8/20/21,Sales,92000,45.0,True
2,103,Peter,Jones,1/2/22,HR,71000,29.0,True
3,104,Mary,Johnson,11/10/23,Engineering,115000,38.0,True
4,105,Mike,Brown,5/30/20,sales,89000,NaN,False


## Task 1: Basic Structure
Show the basic structure of the DataFrame.

Your code should show:
- the shape of the DataFrame;
- the column names;
- the first five rows;
- the last five rows.


In [122]:
# Your code here
print(df.shape)
num_rows, num_columns = df.shape
print(f"This dataset has {num_rows} rows and {num_columns} columns.")
df.columns
print(df.head()) 
print(df.tail())


(13, 8)
This dataset has 13 rows and 8 columns.
  employee_id first_name last_name hire_date   department  salary   age  \
0         101       John     Smith   4/15/22    Marketing   85000  34.0   
1         102       Jane       Doe   8/20/21        Sales   92000  45.0   
2         103      Peter     Jones    1/2/22           HR   71000  29.0   
3         104       Mary   Johnson  11/10/23  Engineering  115000  38.0   
4         105       Mike     Brown   5/30/20        sales   89000   NaN   

  full_time  
0      True  
1      True  
2      True  
3      True  
4     False  
   employee_id first_name last_name hire_date   department    salary    age  \
8          102      Emily     White   1/15/24        Sales     91000   31.0   
9          110      David    Miller   7/22/19           HR     68000   58.0   
10         111      Sarah     Moore   3/12/23    Marketing  $91,000    28.0   
11       error      Kevin     Clark    2/5/24        Sales     88000   35.0   
12         113      Br

### Task 1 Written Response
Write 1–2 sentences describing the size and general structure of the dataset.

**Your response:**  
The dataset has 13 rows and 8 columns. 6 of the 8 rows are strings, age is a float64, and full_time is an object (bool in python). Certain rows contain case-sensitive variables, numbers that can be considered outliers in the sales and age categories, and an error in the employee_id column.

## Task 2: Data Types and Non-Missing Counts
Use Pandas to inspect data types and non-missing counts.

Your code should use:
- `info()`
- `dtypes`


In [132]:
# Your code here
df.info()
df.dtypes

<class 'pandas.DataFrame'>
RangeIndex: 13 entries, 0 to 12
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   employee_id  13 non-null     str    
 1   first_name   13 non-null     str    
 2   last_name    13 non-null     str    
 3   hire_date    12 non-null     str    
 4   department   13 non-null     str    
 5   salary       12 non-null     str    
 6   age          12 non-null     float64
 7   full_time    12 non-null     object 
dtypes: float64(1), object(1), str(6)
memory usage: 964.0+ bytes


employee_id        str
first_name         str
last_name          str
hire_date          str
department         str
salary             str
age            float64
full_time       object
dtype: object

### Task 2 Written Response
Which columns, if any, have data types that might need to be investigated later? For example, is a number or date stored as text/object?

**Your response:**  
Columns 3, 5, 6 & 7 contain data types that might need to be investigated later. Columns 3 & 5 have dates & numbers respectively stored as text, column 6 is a missing number, and 7 is a missing object.

## Task 3: Missing Values
Check missing values for all columns.

Your code should use `isna().sum()`.

Optional: also calculate missing percentages.


In [137]:
# Your code here
print(df.isna().sum())

employee_id    0
first_name     0
last_name      0
hire_date      1
department     0
salary         1
age            1
full_time      1
dtype: int64


### Task 3 Written Response
Which columns have missing values? Which missing values seem most important to investigate later?

**Your response:**  
Columns 3, 5, 6 & 7 have missing values. I think that all missing values are important and should be investigated later, but I would say that hire_date and age are the two I would focus on first because if in the future we want to filter employees by their age or date of hiring then we are going to want to ensure that those values are properly stored in the respective column. I'd say salary is also at the top because we want to make sure every employee is getting paid fairly. 


## Task 4: Numerical Summary
Use `.describe()` to summarize numerical columns.

Then choose one numerical column and calculate at least two individual summary statistics such as mean, median, min, max, or standard deviation.


In [141]:
# Your code here
df.describe()

,age
count,12.000000
mean,118.583333
std,277.411889
min,28.000000
25%,32.500000
50%,36.500000
75%,46.750000
max,999.000000


### Task 4 Written Response
What do you notice from the numerical summaries? Mention at least one range, unusual value, or pattern.

**Your response:**  
After analyzing the values, the maximum value is what caught my attention. A maximum age of 999 is not possible, and this is what is causing all of our other statistics (in particular, the mean and standard deviation) to yield such large numbers. Additionally, we have 13 rows but a count of 12 produced by the missign value we found earlier. 


## Task 5: Categorical Summary
Use `.value_counts()` to summarize at least one categorical column.

Start with `department` if it exists in the dataset. Then choose one additional categorical column if available.


In [182]:
# Your code here
#print(df["employee_id"].value_counts())
#print(df["first_name"].value_counts())
#print(df["last_name"].value_counts())
#print(df["salary"].value_counts())
#I ran the codes above as well, but I noticed that every value appears once (except the employee_id of 102 which appears twice). 
#Instead I wanted to focus on the department, which had counts greater than 1
print(df["department"].value_counts())


department
Sales          4
Marketing      3
Engineering    3
HR             2
sales          1
Name: count, dtype: int64


### Task 5 Written Response
What are the most common categories? Do you notice anything that might need standardization or further investigation later?

**Your response:**  
Sales is the most common cateogory, followed closely by Marketing and Engineering. However, "sales" is counted as an extra category which is due to those case-sensitive variables. 


## Task 6: Duplicate Checks
Check for duplicate records and duplicate employee IDs.

Your code should check:
- fully duplicated rows;
- duplicate `employee_id` values;
- the most common `employee_id` values using `value_counts()`.


In [202]:
# Your code here
row_duplicates = df.duplicated().sum()
print(row_duplicates)
print(f"Number of row duplicates: {row_duplicates}")

employee_id_duplicates = df.employee_id.duplicated().sum()
print(employee_id_duplicates)
print(f"Number of employee ids duplicates: {employee_id_duplicates}")
df["employee_id"].value_counts().head(10)

0
Number of row duplicates: 0
1
Number of employee ids duplicates: 1


employee_id
102    2
101    1
103    1
104    1
105    1
106    1
107    1
108    1
110    1
111    1
Name: count, dtype: int64

### Task 6 Written Response
Are there duplicate employee IDs? If yes, what might that mean if `employee_id` is supposed to uniquely identify each employee?

**Your response:**  
The id of 102 is duplicated. If employee_id is duplicated we might have an issue if we are trying to filter employees by their id. This would be like having one input yielding two outputs, so it might cause a potential error in the future. Additionally, it could also give a false count of employees if we are trying to cound them by their id.

## Task 7: Overall Data Quality Observations
Write 3–5 bullet points summarizing the condition of the dataset.

Focus on what you found from profiling. Do **not** clean the data yet.

Examples:
- The dataset has ___ rows and ___ columns.
- The ___ column has missing values.
- The ___ column may have an incorrect data type.
- The ___ column has possible duplicate IDs.
- The ___ column has a range from ___ to ___.


**Your observations:**

1. The dataset has 13 rows and 8 columns.
2. The hire_date, salary, age and full_time columns have missing values.
3. The salary column may have an incorrect data type.
4. The employee_id column has possible duplicate IDs.
5. The age column has a range from 28.0 to 999.0.


## Submission Checklist
Before submitting, make sure:

- [ ] All code cells run without errors.
- [ ] You completed all tasks.
- [ ] You wrote short responses in the markdown cells.
- [ ] You saved the notebook.
- [ ] You committed and pushed your work to GitHub.

Use the terminal:

```bash
git add .
git commit -m "Completed Week 2 data profile checkpoint"
git push
```
